README.md
echoping
========

echoping **was** a small program to test (approximatively) performances of a remote host by sending it requests such as HTTP requests.

echoping is **no longer** maintained.

To use it, simply:

```
% echoping machine.somewhere.org
```

or use the options before the machine name (see the man page).

See the DETAILS file for various traps when benchmarking networks, specially with this program.

In any case, be polite: don't bother the remote host with many repeated requests, especially with large size. Ask for permission if you often test hosts which aren't yours.

Current features:

    * plugins, so you can extend echoping with any protocol you like and/or use,
    * Supports IPv6 as well as IPv4,
    * Supports IDN (Unicode domain names like café.gennic.net),
    * uses the protocols echo, discard, chargen or HTTP,
    * can use cryptographic connections with HTTP,
    * uses UDP instead of TCP for the protocols which accept it (like echo),
    * can repeat the test and display various measures about it, 

Examples of output:

```
    (Simple test with 1000 bytes echo TCP packets)

    % echoping -v -s 1000 mycisco

    This is echoping, version 5.0.0.

    Trying to connect to internet address 172.21.0.14 7 to transmit 1000 bytes...
    Connected...
    TCP Latency: 0.003165 seconds
    Sent (1000 bytes)...
    Application Latency: 0.322183 seconds
    1000 bytes read from server.
    Checked
    Elapsed time: 0.326960 seconds

    (Repeated tests with average / mean and median displayed.)

    % echoping -n 10 faraway-machine
    [...]
    Minimum time: 6.722336 seconds (38 bytes per sec.)
    Maximum time: 17.975060 seconds (14 bytes per sec.)
    Average time: 10.873267 seconds (24 bytes per sec.)
    Standard deviation: 3.102793
    Median  time: 9.218506 seconds (28 bytes per sec.)

    (Testing a Web server with an HTTP request for its home page.)

    %  echoping -h / mywww
    Elapsed time: 0.686792 seconds
```

The exit status is set if there is any problem, so you can use echoping to test repeatedly a Web server, to be sure it runs fine (SmokePing does it).

SRC/ChangeLog
6.0.2 (2007-04-09)

	* Bug fixes only
	
6.0.1 (2007-04-04)

	* Bug fixes only
	
6.0.0 (2007-03-26)

	* Plugins: several plugins (whois, postgresql, dns) are now
	shipped with echoping and you can write your own. All new
	protocols will now be supported via plugins. Consequences:
	   * you now need an Unix with a working dlopen(3)
	   * echoping now requires the popt parsing library
	* popt: echoping now depends on popt. One main change: options
	parsing is lightly different so some scripts *may* break. A
	typical example is that "old-school" Unix aggreation of arguments
	("echoping -v4 localhost") is no longer accepted (use "echoping -v
	-4 localhost").
        * Many bug fixes, see the list at Sourceforge
	
5.2.0 (2004-02-29)

	* IDN (Unicode domain names) support
	
5.1.0 (2003-11-09)

	* GNU TLS support (alternative to OpenSSL),
	* Small licence change,
	* Portability fixes (MacOS X)
	
5.0.1 (2002-11-04)

	* Small bug fixes,
	* Better error messages with invalid options

5.0.0 (2002-10-11)

	* Now uses getnameinfo/getaddrinfo. Two consequences:
	   - you can compile only on a recent Unix
	   - you have IPv6 support
        * With -v, now able to separate kernel (TCP) and application latency
	* Now displays the standard deviation for repeated tests. Warning:
	  this may break some scripts which parse the output of echoping.
	
4.2.0 (2001-03-16)
	
        * New -a and -A options to control the proxy/cache (disable caching).
	  Fixes bug/wish #232324
	* Support for microsecond wait (uses usleep)
	* New option (-p) to set socket priority (Linux only)
	* New option (-P) to set IP Type of Service octet
	* Bug fixes: #230384 (SSL records are different from text lines)

4.1.0 (2001-02-14)

	* Fix several SSL bugs (#132001, #130151)
	* Now requires an ANSI C compiler
	* Now uses only stdarg.h so should really compile everywhere

4.0.2 (2001-01-26)

	* Stupid bug introduced in 4.0.1 when compiling without SSL

4.0.1 (2001-01-26)

	* Fix SSL bug when using -n (#130078)

4.0.0 (2001-01-25)

	* SSL (Secure Sockets Layer) support
	* Now reading many bytes at a time (big performance improvment).
	  Uses the standard I/O library.
	* Better detection of the HTTP port
	* Bug fixes (AI/X, Solaris)
	
3.1.0 (never released)

	* SMTP support
	
3.0.1 (2000-07-06)

	* configure now tries to detect the name of the HTTP port
	* Fix an AIX bug (closes #108782)

3.0.0 (2000-06-30)

	* Autoconf/automake support
	* Fix a bug in ICP timeout
	  
2.2.1 
        * Portability fixes

2.2
        * Stable release, apart from ICP

2.1.0
	* Timeouts even on TCP connections.
	* Support of HTTP 1.1
	
2.0.1

	* Bug with getservbyname for HTTP connections fixed.
	
2.0.0 
	* HTTP support. For testing of Web servers and Web proxies.
	* Much better portability.
   

$Id$






	

SRC/icp.h
#ifndef ICP_HEADER

#define ICP_HEADER

/* Version */
#define ICP_VERSION_1           1   
#define ICP_VERSION_2           2
#define ICP_VERSION_3           3
#define ICP_VERSION_CURRENT     ICP_VERSION_2

#define ICP_FLAG_HIT_OBJ	0x80000000ul
#define ICP_FLAG_SRC_RTT	0x40000000ul

#if SIZEOF_INT == 4
typedef unsigned int u_num32;
#elif SIZEOF_LONG == 4
typedef unsigned long u_num32;
#else
#error "No suitable type for representing a 32-bits value"
#endif

struct icp_common_s {
    unsigned char opcode;       /* opcode */
    unsigned char version;      /* version number */
    unsigned short length;      /* total length (bytes) */
    u_num32 reqnum;             /* req number (req'd for UDP) */
    u_num32 flags;
    u_num32 pad;
    u_num32 shostid;            /* sender host id */
};
typedef struct icp_common_s icp_common_t;

typedef enum {
    ICP_OP_INVALID,		/* 00 to insure 0 doesn't get accidently interpreted. */
    ICP_OP_QUERY,		/* 01 query opcode (cl->sv) */
    ICP_OP_HIT,			/* 02 hit (cl<-sv) */
    ICP_OP_MISS,		/* 03 miss (cl<-sv) */
    ICP_OP_ERR,			/* 04 error (cl<-sv) */
    ICP_OP_SEND,		/* 05 send object non-auth (cl->sv) */
    ICP_OP_SENDA,		/* 06 send object authoritative (cl->sv) */
    ICP_OP_DATABEG,		/* 07 first data, but not last (sv<-cl) */
    ICP_OP_DATA,		/* 08 data middle of stream (sv<-cl) */
    ICP_OP_DATAEND,		/* 09 last data (sv<-cl) */
    ICP_OP_SECHO,		/* 10 echo from source (sv<-os) */
    ICP_OP_DECHO,		/* 11 echo from dumb cache (sv<-dc) */
    ICP_OP_UNUSED0,		/* 12 */
    ICP_OP_UNUSED1,		/* 13 */
    ICP_OP_UNUSED2,		/* 14 */
    ICP_OP_UNUSED3,		/* 15 */
    ICP_OP_UNUSED4,		/* 16 */
    ICP_OP_UNUSED5,		/* 17 */
    ICP_OP_UNUSED6,		/* 18 */
    ICP_OP_UNUSED7,		/* 19 */
    ICP_OP_UNUSED8,		/* 20 */
    ICP_OP_MISS_NOFETCH,	/* 21 access denied while reloading */
    ICP_OP_DENIED,		/* 22 access denied (cl<-sv) */
    ICP_OP_HIT_OBJ,		/* 23 hit with object data (cl<-sv) */
    ICP_OP_END			/* 24 marks end of opcodes */
} icp_opcode;


#endif  /* ICP_HEADER */
